### PR TITLE
[4.0] Update PrefixField.php in Installer

### PR DESCRIPTION
**Pull request by Duke3D**
Pull Request for Issue # .
First report.

### Summary of Changes
Added to validation of user-entered database prefix.
1) user's response is converted to lowercase so that value stored in configuration.php is in alignment with the default behavior (convert to lowercase) of the MySQL database.  
2) test for and append underscore, if missing.

### Testing Instructions
Run Installer.  
A)  At Database Prefix field, enter mixed or upper case prefix ending with underscore.  
B)  Repeat except enter mixed or upper case prefix without underscore.
 
### Expected result
Installed database has lowercase prefix and value matches database prefix as stored in configuration.php.

### Actual result
If not patched, installed database has lowercase prefix.   Database prefix stored in configuration.php has mixed case prefix.  Can lead to database issues on Linux servers and reference errors that appear on server migration.

### Documentation Changes Required
None

### CODE REVIEW

**Comment by infograf768** [comment]
This deals only for the session prefix. It should also deal with new prefix when modified, therefore out of the if/else.
Also cs is bad.

**Comment by brianteeman** [comment]
To satisfy the codestyle requirements please address the following issues

```
FILE: /********/src/installation/src/Form/Field/Installation/PrefixField.php
--
46 | --------------------------------------------------------------------------------
47 | FOUND 4 ERROR(S) AFFECTING 3 LINE(S)
48 | --------------------------------------------------------------------------------
49 | 82 \| ERROR \| Please consider a blank line preceding your comment
50 | 85 \| ERROR \| Please consider a blank line preceding your comment
51 | 86 \| ERROR \| Expected "if (...)\n"; found "if (...) "
52 | 86 \| ERROR \| Closing brace must be on a line by itself


```

File: installation/src/Form/Field/Installation/PrefixField.php [hunk 1]
```diff
@@ -79,6 +79,14 @@ protected function getInput()
 		else
 		{
 			$prefix = $session['db_prefix'];
+
```
Comment (Quy) [inline installation/src/Form/Field/Installation/PrefixField.php]: Remove tabs.

File: installation/src/Form/Field/Installation/PrefixField.php [hunk 2]
```diff
@@ -79,6 +79,14 @@ protected function getInput()
 		else
 		{
 			$prefix = $session['db_prefix'];
+			
+			// Ensure user-provided prefix is lowercase
+			$prefix = strtolower($prefix);
+
```
Comment (Quy) [inline installation/src/Form/Field/Installation/PrefixField.php]: Remove tabs.

File: installation/src/Form/Field/Installation/PrefixField.php [hunk 3]
```diff
@@ -79,6 +79,14 @@ protected function getInput()
 		else
 		{
 			$prefix = $session['db_prefix'];
+			
+			// Ensure user-provided prefix is lowercase
+			$prefix = strtolower($prefix);
+			
+			// If last character from user is not an underscore then append the underscore.
+			if (strpos($prefix, '_') === false) {
```
Comment (Quy) [inline installation/src/Form/Field/Installation/PrefixField.php]:
```suggestion
			if (strpos($prefix, '_') === false)
			{
```

Comment (Duke3D) [reply installation/src/Form/Field/Installation/PrefixField.php]: Thank you

**Comment by Quy** [comment]
@Duke3D Please do as requested here: https://github.com/joomla/joomla-cms/pull/25144#issuecomment-500194741

**Comment by Schmidie64** [comment]
I have tested this item :red_circle: unsuccessfully on 0e7b5cadc90740e78fc1fd0160ff8fe67298212b<br /><br />After applying the patch, the prefix in the database is in lower case (without the patch  it's also in lower case). In the configuration.php is the prefix still in mixed cases. So the pr doesen't work well.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25144">issues.joomla.org/tracker/joomla-cms/25144</a>.</sub>

**Comment by richard67** [comment]
To fix https://github.com/joomla/joomla-cms/pull/25144#issuecomment-500194741 , the checks have to be added to 2 places:
- https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/src/Model/SetupModel.php#L313
- https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/src/Model/DatabaseModel.php#L95

(Or maybe just the regex needs to be modified).

If my PR #29362 will be ready and merged, it will be only one place where to change that.

But I am in doubt if this PR here is right, because we had such check already for PostgreSQL, where it seems to matter, e.g. here:
https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/src/Model/DatabaseModel.php#L133

So I think it could have had a reason why we only did it for PostgreSQL and not for MySQL.

But who knows, maybe it was wrong all the time for MySQL.

**Comment by roland-d** [comment]
@Duke3D  Any interest in picking this up or should the PR be closed?

**Comment by Quy** [comment]
Closing due to no response.

**Pull request closed without merging**